### ISSUE TITLE
Support waiting for an UI to attach when launching with --listen

### ISSUE THREAD

**Issue by fredizzimo** [neovim/neovim#29634]
### Problem

Currently you can launch a remote server with `--headless --listen addr:port`, and later connect an UI like Neovide to it, which is good. But the problem is that Neovim does not wait for the UI to attach and proceeds to load the init files immediately. This is problematic because the user might have conditional customization in the init files, for example `if vim.g.neovide`, which will never run.

We, the maintainers of Neovide, have always assumed that `--headless` waits for the UI to attach, so that's what we have been suggesting in the documentation. But as part of investigating bugs like https://github.com/neovide/neovide/issues/2620, https://github.com/neovide/neovide/issues/1868 I found that to not be the case. It's also not a bug, since the behaviour is clearly documented in the documentation (but not the commandline help).

https://github.com/neovim/neovim/pull/25709 (Neovim 0.10) added support for `--embed --listen addr:port`, which waits for the UI to attach. But that is only suitable as a workaround at best, since it also captures the stdin. Another hacky workaround is to do the GUI specific configuration in `UIEnter`, but that's too late due to https://github.com/neovim/neovim/issues/25377, and things like font's should definitely be set before anything is drawn on the screen.

### Expected behavior

Preferably `--headless --listen addr:port`, should always wait for an UI to attach, at least I can't think of any use cases for the current behavour. 

But if that's not possible, a new command line parameter should be added to make the startup wait for an UI to attach.

**Comment by glacambre** [comment]
The approach I took in Firenvim is to add a bunch of `--cmd` switches to perform UI-specifc initialization. Would having Neovide add `--cmd 'source /path/to/neovide-init.lua'` to the arguments it passes to Neovim be an acceptable solution?

**Comment by cosmicboots** [comment]
@glacambre using `--cmd` flags might work, but it feels slightly like a hack in this case, and I'm not sure it's the most general solution.

I can't speak on behalf of the neovide maintainers, but I also don't think sourcing a custom startup file would work in all cases. Currently, neovide allows you to customize _any_ section of the config by simply checking if the global `vim.g.neovide` variable is set. This allows both adding configuration for the GUI, and preventing terminal specific configurations from running.

Using `--cmd` flags helps out with adding configuration when launching the GUI, but it doesn't handle preventing other code from running.

**Comment by glacambre** [comment]
@cosmicboots I'm not sure I understand what you mean. I'm suggesting adding `--cmd 'let g:neovide = true'` to the arguments Neovide passes to nvim. As the [docs](https://neovim.io/doc/user/starting.html#--cmd) explain, this piece of code will be executed before any vimrc/init.vim is executed - thus `g:neovide` would be defined when the vimrc is executed, and users could use `if exists('g:neovide') && g:neovide` in order to gate their configuration. Gating is even nicer in lua since undefined variables will be nil, which is falsy.

**Comment by fredizzimo** [comment]
@glacambre,

The use case here is to launch a remote Neovim instance, which you can later connect to using `neovide --server host:port`

It feels quite hacky to let the user start the instance with `nvim --headless --listen --cmd 'let g:neovide = true`, since neovide is supposed to be set by Neovide itself and not by the user.

And it does also not cover all the cases. Right now, it's mostly OK; the checks in the user configuration are not really needed, since Neovide is configured using global variables (https://neovide.dev/configuration.html), so they can be set no matter if Neovide is connected or not. They are typically only needed for doing something else when Neovide is *not* running, or for configuring standard Neovim options based on the configuration.

But we probably want to add custom commands and functions that can be run during the configuration stage at some point, perhaps a traditional `setup` function or something else, since the flat variable list is not really extendible enough for what we have been planning. Some options were discussed here https://github.com/neovim/neovim/issues/12544, and they all require that Neovide, or at least some Neovide plugin is running when the init file is parsed. And having a separate plugin that needs to be in sync with Neovide sounds like a nightmare for the users.

So as far as I see it, the absolutely cleanest solution is to have a parameter that waits at the startup, just like `--embed` does. That way the GUI can add its own configuration before the user init file is parsed.

**Comment by glacambre** [comment]
Ah, okay, I understand better now. On a philosophical level, I'm against this proposal - as you said in the OP, there's already `UIEnter` to do this, and it's strictly better than setting things in init.vim/init.lua based on whether a global variable is defined, since it'll work when multiple GUIs connect to a same Neovim instance or when a same UI connects and disconnects to the same instance multiple times.

However, on a practical level, I must admit that in my experience users clearly don't understand `UIEnter` and don't want to learn anything more than global variables. Moreover, aside from combining noice.nvim and another GUI, there has been very little use of combined GUIs, despite this being possible since the inception of Neovim's UI protocol. Finally, I must admit that having a better way to delay sourcing of the user's init.vim/init.lua would likely have helped me a lot with Firenvim. So in the end, I'd say this change would indeed be a very worthwhile improvement to Neovim.

Yet another case of worse is better I guess :)

**Comment by bfredl** [comment]
`--headless` kinda implies that an UI is _not_ expected (cf `--embed` with `--embed --headless`) so if we support this specific pattern a different flag would be preferable.

**Comment by fredizzimo** [comment]
@glacambre,

`UIEnter` is not really an option since it's not guaranteed to be called before the UI is rendered. For example, `lazy.nvim` draws its UI completely before `UIEnter`. Many error messages are also displayed before that. So, if you set for example the `guifont` there, the initial window size and font will be wrong. See https://github.com/neovim/neovim/issues/25377

So, for this reason in Neovide, we assume that all the options are valid when the screen is drawn for the first time and the first flush message is received. We delay showing our UI window until that point.

**Comment by justinmk** [comment]
>> Preferably `--headless --listen addr:port`, should always wait for an UI to attach,
>
> `--headless` kinda implies that an UI is _not_ expected

Should we add a flag that sets a mode where Nvim just doesn't proceed past "x" until some sort of global function `startup()` is called? Then when a UI, or whatever else, connects, it can call `startup()` when it's ready. That avoids needing more flags for the various permutations of scenarios. Or maybe there aren't more scenarios after this one? :)

**Comment by fredizzimo** [comment]
`--clean --headless --listen addr:port` could work, if we have an api function to trigger the configuration loading using the correct runtime paths, takiing everything into account, like `NVIM_APP` set on the remote for example. But it also needs an extra argument to `neovide` to enable this behavior, we don't know if the remote server has loaded the configuration or not, without the user telling us.

Another option is to fix `UIEnter`, so that can be used for the UI specific configuration.

Currently, that's not a viable option for Neovide in the general case, since the screen is drawn before that using the wrong configuration  like wrong fonts, causing redraws and flickering.

* https://github.com/neovim/neovim/issues/25377

**Comment by justinmk** [comment]
> But it also needs an extra argument to `neovide` to enable this behavior, we don't know if the remote server has loaded the configuration or not, without the user telling us.

If the server wasn't started by the UI (e.g. neovide) then of course it wouldn't wait for the UI. I don't see how that's relevant here. And even if it did, another UI may connect later...




> Another option is to fix `UIEnter`

Let's start with that. I reviewed https://github.com/neovim/neovim/issues/25377 again, and I don't see what is blocking us from redefining the behavior of UIEnter. See https://github.com/neovim/neovim/issues/25377#issuecomment-2355084760

**Comment by fredizzimo** [comment]
> If the server wasn't started by the UI (e.g. neovide) then of course it wouldn't wait for the UI. I don't see how that's relevant here. And even if it did, another UI may connect later...

That's the problem here, Neovide can't start the server on a remote machine, just connect to it. So, we need a way to make that server wait for the connection.

And, even if we added ssh support for example, it would still not cover any case where you login to the remote some other way to start it.

**Comment by justinmk** [comment]
> That's the problem here, Neovide can't start the server on a remote machine, just connect to it. So, we need a way to make that server wait for the connection.

That case, by definition, happens after `VimEnter`, and indeed the server may be a long-lived session. Why would it need to wait for a UI to attach? And what if another UI already attached?

*Assume* that a UI *already* attached, and now Neovide (a second, "incoming" UI) also wants to attach. The incoming UI may set `:help ui-override`, and may also override various other editor options that affect UI. What's missing from that scenario? Let's focus on that and solve that scenario.

**Comment by justinmk** [comment]
> #25709 (Neovim 0.10) added support for `--embed --listen addr:port`, which waits for the UI to attach.

We could make `--remote-ui --listen` wait for a UI. How about that? cc @bfredl

**Comment by justinmk** [comment]
> The user has something like this in the configuration, but it does not get applied due to `vim.g.neovide` not being set.

But Neovide didn't start the server, so why would anyone expect neovide globals to be set during startup? Of course a `UIEnter` is required to handle when a UI attaches.

**Comment by clason** [comment]
> The dictwatcher approach that we are currently using only supports watching at the top level

Which is a limitation of the dictwatcher that should be fixed, no? Instead of worked around in increasingly byzantine constructions?

**Comment by justinmk** [comment]
I don't think we should enhance dictwatcher, but https://github.com/neovim/neovim/issues/12544#issuecomment-1116794687 is on the table.

**Comment by clason** [comment]
Well, whether it's by improving the builtin dictwatcher or adding a new Lua-only "tablewatcher" is secondary here; the point is that there's missing functionality that a) seems broadly desirable and b) would obviate a lot of the discussion here.

**Comment by clason** [comment]
Pro tip: In these issues, focus on the Neovide side (i.e., specify exactly what _you_ need from _your_ side, independent of how you are _currently_ trying to achieve that) and leave the discussion on how _Neovim_ should implement to later/us.

**Comment by justinmk** [comment]
@fredizzimo related to this, I recently noticed that there doesn't seem to be a way to "detach" from a `nvim --embed` process: closing the channel from either end causes nvim to exit? I can definitely see a use-case there (also), would that be useful to you?

**Comment by fredizzimo** [comment]
I don't personally see much need for it, but that doesn't mean that there aren't any. 

I think it could address the request, although I don't think it would be as ergonomic as a dedicated argument. I think you either have to run some additional commands after nvim starts, or somehow pipe stdio over the network. If you have an ssh connection, that's easy but not every remote machine has that, and then you need to use something like socat.

**Comment by justinmk** [comment]
> although I don't think it would be as ergonomic as a dedicated argument.

For the socket (not TCP) use-case, `--embed` has nice ergonomics because it lets Nvim choose the listen path. Then once the caller gets the `v:servername`, it could detach and continue with the socket connection instead of stdio.
 
> I think you either have to run some additional commands after nvim starts, or somehow pipe stdio over the network

Wouldn't it also work to run the `nvim --embed --listen 127.0.0.1:x` script remotely, which allows the UI to attach to the TCP `--listen` address and then close the stdio channel?

**Comment by fredizzimo** [comment]
> For the socket (not TCP) use-case, --embed has nice ergonomics because it lets Nvim choose the socket path. Then once the caller gets the v:servername, it could detach and continue with the socket connection instead of stdio.

But then you need to attach to it through stdio first or pass some additional command line parameters to run some scripts, or am I misunderstanding what you mean here.

> Instead of piping stdio, run nvim --embed --listen 127.0.0.1:x script remotely, which waits until a UI attaches to the TCP --listen address, which then can close the stdio channel. Any problems with that?

`nvim --embed --listen` already does almost exactly what's needed, except that it also listens to stdin for connections. This is problematic, since you can type something in the terminal by mistake. 

But now after reading https://github.com/neovim/neovim/pull/25709, and https://github.com/neovim/neovim/issues/25706, I don't think that was ever part of the requirement, but a side effect of the implementation.  So, we could fix this issue by simply redefining it to always wait for only the remote connection.

**Comment by Glirastes** [comment]
@fredizzimo I don't think having `--listen` wait for the first UI attach is a good idea. Generally, I think this "waiting for the first UI attach" is non standard behaviour and should have to be opt-in. `--listen` does exactly one thing and that is to enable connecting to that specified address later. I think it should stay that way.
Though I am absolutely with you that there should be a way to start nvim without it's own TUI, listening to a socket, waiting for the first UI attach and not capturing stdin.

@justinmk 
The problem I see with `--embed` and `--headless` is that they already touch too many different things: They prevent nvim from starting it's own UI, they capture stdio and (in the case of `--embed`) they enable this deferred initialization.
To me, `--headless` implies exactly one thing: "start without your own TUI". I wouldn't expect it to open a stdio channel. In fact, I would expect *exactly that* to be the difference between `--headless` and `--embed`. If I embed nvim inside my application, nvim should first of all not render it's own UI (like `--headless`), and since it sits *inside* my application, I (can) communicate with it via stdio.
Currently the defining difference is another: `--embed` waits for the UI to attach, while `--headless` does not. They differ in the type of stdio channel they open as well.
To me, embedding nvim does not imply having it wait for the first UI attach. Sure, many UIs might want this behavior, but others may want nvim to start and init normally and handle their stuff in UIEnter. I consider this "waiting for UI attach" an implementation detail which shouldn't be the defining characteristic of "embedding nvim".
Maybe that's just my interpretation of what `--headless` and `--embed` imply semantically.

I'm thinking about the following:
1. Have `--headless` not open a channel on stdio, but *just* make nvim start without it's TUI
2. Have `--embed` not wait for the first UI to attach, but rather behave like `--headless` with the addition that it opens an RPC channel on stdio
3. Introduce a new switch for having nvim wait for the first UI to attach

The lost functionality of `--headless` could be regained by opening a channel through `--cmd`.
The lost functionality of `--embed` would be provided by 3., something like `--wait-for-ui`. This would make this behaviour more explicit, but at least allow one to get this behaviour *without* having stdin captured automatically.
Summarizing, I think the problem is that `--embed` and `--headless` do too much at once and should be broken apart. There are desirable combinations of behaviour that should involve them, but you're buying into too much at once.

**Comment by justinmk** [comment]
I hide that last comment because it didn't address https://github.com/neovim/neovim/issues/29634#issuecomment-2422234572 . I am proposing:

    nvim --embed --listen foo -c 'call chanclose(1)'

That will leave the server in a "waiting for UI" state, and it addresses this issue.

**Comment by fredizzimo** [comment]
From the perspective of the user I'm not sure what `call chanclose(1)` would do there if you never attach anything else than the remote UI. I know that it internally opens a channel, but it doesn't seem like implementation details that should be exposed and documented.

The command is also fairly hard to remember when you login to a remote server and want to use it.

**Comment by fredizzimo** [comment]
It also does not work, at least not with the current master version. It still keeps the stdin open until the remote UI tries to attach, so pressing enter for example exits with an error. And as soon as the remote UI attaches, it exits with the following
```
ERROR [neovide::bridge::ui_commands] Resize failed

Caused by:
    0: Error sending request 'nvim_ui_try_resize'
    1: Error writing to the Writer
    2: Broken pipe (os error 32)
```

**Comment by fredizzimo** [comment]
I think that https://github.com/neovim/neovim/issues/25377, is the more generic way of resolving the issue. Especially with the new developments in regard to dynamically attaching and detaching different UI. That way the UI will always have a chance to configure itself before anything is rendered with that UI, this is especially important if a different type of UI is reconnected. And the command-line option would not help with resolving that problem, it only helps when the first UI is connected.

**Comment by justinmk** [comment]
> I'm not sure what `call chanclose(1)` would do there if you never attach anything else than the remote UI.

It waits for a UI to attach, as `nvim --embed` always does, and as the alternatives proposed here would?

> The command is also fairly hard to remember when you login to a remote server and want to use it.

Why would a *user* start a remote `nvim` that "waits for UI"?? I understood the use-case as "a GUI wants to start a remote Nvim for the user".

> It also does not work, at least not with the current master version.

That is why this issue is still open. But it will work pretty soon, only a small change is needed.

> I think that [#25377](https://github.com/neovim/neovim/issues/25377), is the more generic way of resolving the issue.

That proposal is essentially undefined beyond "rework the redraw order somehow".

**Comment by fredizzimo** [comment]
> Why would a user start a remote nvim that "waits for UI"? I understood the use-case as "a local GUI wants to start a remote Nvim for the user".

Because they are working on a remote server and want to connect Neovide to it, that's the primary use case for this. If there's no server involved, then the standard `--embed` works just fine.

Neovide could maybe support *one* type of remote, ssh ones by utilizing some library, but I think that's out of scope for it. And there are other types of remotes, where you might only have a remote desktop connection, or even a web based shell.

> That proposal is essentially undefined beyond "rework the redraw order somehow".

I don't think so, the only requirement is that `UIEnter` (if added is called before anything is drawn using that UI). The discussion there got out of track because of too many different other proposals.

**Comment by justinmk** [comment]
Ok so to address the original motivation in this issue description...

# Proposal

> the user might have conditional customization in the init files, for example `if vim.g.neovide`, which will never run.

...we have two clear paths (and both are planned/feasible in the short term):

1. Users that want to check `g:neovide` at startup, can use the `chanclose()` approach. 
    - We aren't going to add random startup flags for that use-case.
3. Users that define a `UIEnter` handler can start `nvim` normally on their remote, then `:detach`, then connect with a GUI.
    - Depends on [#25377](https://github.com/neovim/neovim/issues/25377) so that `UIEnter` is predictable/reliable.

**Comment by justinmk** [comment]
# Proposal (additional)

> `nvim --embed --listen` already does almost exactly what's needed, except that it also listens to stdin for connections. This is problematic, since you can type something in the terminal by mistake.

How about this: When user invokes `nvim --embed` in a terminal, Nvim auto-daemonizes; does *not* connect to stdin; and immediately exits the builtin UI. The `nvim --embed` server continues running as a detached (background) process, waiting for UI connect before continuing startup.

**Comment by fredizzimo** [comment]
Yes, that would work to fix the immediate problem and is almost the same as I suggested here https://github.com/neovim/neovim/issues/29634#issuecomment-2425831446. The difference is that that I'm suggesting a blocking implementation, just like `nvim ---headless -listen foo` instead of running in the background.

It does not solve the future use cases, where you might start the TUI first, then realize that you want to work in the GUI, or when you connect to multiple GUIs and allow them to be configured differently. Personally, I would rather see the `UIEnter` approach implemented, and I might even work on it at some point in the not so distant future. That also means that we can move over to a more powerful lua based way of configuring Neovide. We could do it before with your proposal, but that means two big configuration breaking changes for Neovide instead of one.

Also with the current configuration, we have a satisfactory workaround, everything can be configured out of order, and it doesn't matter if Neovide is connected or not because we use just global variables. So, the only thing users need to do is to remove all the `if vim.g.neovide` checks

**Comment by mikew** [comment]
Not to drudge up an old issue but I'm facing similar and it's no clear why `--embed` works the way it does. Issue reference for background: https://github.com/mikew/nvrh/issues/79

My tool runs neovim on a server via SSH, tunnels the socket to your local machine, and starts neovim locally talking over the socket. I'm using `--headless`, but someone is asking for `--embed` (also a Neovide user) so that the server instance waits for a client to connect.

My surprises as a passer-by are:

- It's not clear why `--embed` waits for a UI to attach from the name
- It's not clear why `--embed` fails if there's no stdin (the neovim server instance in my setup does not have stdin available to it)
- [Why the proposed `--embed`](https://github.com/neovim/neovim/issues/29634#issuecomment-2772837520) would switch to daemonize?
- Why a `--wait-for-first-ui` isn't an option

**Comment by justinmk** [comment]
>  It's not clear why `--embed` waits for a UI to attach from the name

because without a UI, user config may show dialogs or do other random things.

> It's not clear why `--embed` fails if there's no stdin

how else can rpc work? maybe if `--listen` is specified? that might be a good idea.

>  Why a `--wait-for-first-ui` isn't an option

of course we can always add more flags, but (1) ideally the existing flags should make sense and be intuitive, and (2) more flags means more permutations of flags.  so "add more flags" is not the first choice.

**Comment by mikew** [comment]
> > It's not clear why --embed fails if there's no stdin
>
> how else can rpc work? maybe if --listen is specified? that might be a good idea.

Yeah in my case I'm trying `--embed --listen ...`, so it's a little bit of a surprise it requires stdin.

> > Why a --wait-for-first-ui isn't an option
>
> of course we can always add more flags, but (1) ideally the existing flags should make sense and be intuitive, and (2) more flags means more permutations of flags. so "add more flags" is not the first choice.

For sure, I tried to frame my post as someone without much context into neovim internals. From that point of view "`--embed` is the same as `--headless` but also waits for the first UI to attach" isn't obviously better than something like `--headless --wait-for-ui`. The technical goals from neovim's POV seem to be:

- Optionally start UI
- Optionally wait for UI to attach
- Optionally require stdin

Trying to cram those 3 booleans into two CLI flags seems like a tall order.

**Comment by justinmk** [comment]
but as you mentioned, there isnt necessarily a reason that `--embed` needs to fail, so a new flag is just more surface area. rather we can make `--embed` work better.

**Comment by mikew** [comment]
Absolutely, I'm just overstepping my bounds wondering about how to expose functionality :)
